### ISSUE TITLE
code compiles error in D3 with typescript v2.5.x

### ISSUE THREAD

**Issue by musicq** [Microsoft/TypeScript#18406]
**TypeScript Version:**  2.5.x

**Code**

```ts
// snippet
  return s.length < 2 ? (q[0]
    ? one(q[0].x)
    : zero(b))
    : (b = q.length, function(t) {
        for (var i = 0, o; i < b; ++i) s[(o = q[i]).i] = o.x(t);
        return s.join("");
      });
  };
```

**Expected behavior:**
Compile correc.
**Actual behavior:**
Both in javascript and typescript, it reports error when compile

```log
ERROR in [at-loader] ./apps/home/js/lib/d3.js:2567:25
    TS1005: '=' expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2570:7
    TS8010: 'types' can only be used in a .ts file.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2570:22
    TS1138: Parameter declaration expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2570:34
    TS1005: '=>' expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2573:8
    TS1005: ':' expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2573:9
    TS1005: ')' expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:2574:1
    TS1128: Declaration or statement expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:16405:1
    TS1128: Declaration or statement expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:16405:2
    TS1128: Declaration or statement expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:16405:3
    TS1128: Declaration or statement expected.

ERROR in [at-loader] ./apps/home/js/lib/d3.js:16405:4
    TS1128: Declaration or statement expected.
```

This happens only in v2.5.x, with 2.4.x, it works fine.

**Comment by mhegazy** [comment]
@sandersn looks like another parsing issue, let's see if we can get this into 2.5.3

**Comment by sandersn** [comment]
Fixed by #18294

**Comment by sandersn** [comment]
Here's a smaller repro that compiles without errors on 2.4.x:

```ts
function foo(q: string, b: number) {
    return true ? (q ? true : false) : (b = q.length, function() { });
};
```

**Comment by sandersn** [comment]
#18417 is a better fix than #18294 that works by requiring speculatively parsed arrow functions to have an `=` for their parameter's initalisers. Currently `(x y) { }` is parsed as if the user intended to write `(x = y) => { }`. This is true sometimes, but often it is not, especially when we parse `y` incorrectly.
